requests_oauthlib/oauth2_session.py

```diff
@@ -334,13 +351,42 @@ def request(self, method, url, data=None, headers=None, withhold_token=False,
                     log.debug('Auto refresh is set, attempting to refresh at %s.',
                               self.auto_refresh_url)
 
-                    # We mustn't pass auth twice.
-                    auth = kwargs.pop('auth', None)
-                    if client_id and client_secret and (auth is None):
+                    # Someday this code can be eliminated, we should be able to
+                    # simply pass **self.auto_refresh_kwargs to
+                    # `refresh_token()` what follows is to maintain
+                    # compatibility
+
+                    # Start with kwargs explicitly requested for refresh.
+                    refresh_kwargs = self.auto_refresh_kwargs.copy()
+
+                    if 'auth' in kwargs:
+                        # If user supplied `auth` to `request()` warn them to
+                        # instead use `auto_refresh_kwargs`. But honor their
+                        # intent for now.
+                        warnings.warn('auth argument supplied. Please specify '
+                                      'token refresh authentication in '
+                                      'auto_refresh_kwargs.')
+                        refresh_kwargs['auth'] = kwargs.pop('auth')
+                    elif client_id:
+                        # If no auth was provided, but `client_id` and
+                        # `client_secret` were, warn the user and create an
+                        # HTTP Basic auth header. It is preferred if the user
+                        # instead place an auth param into `auto_refresh_kwargs`
+                        warnings.warn('client_id argument supplied, Please '
+                                      'specify token refresh authentication in'
+                                      'auto_refresh_kwargs')
                         log.debug('Encoding client_id "%s" with client_secret as Basic auth credentials.', client_id)
-                        auth = requests.auth.HTTPBasicAuth(client_id, client_secret)
+                        if not client_secret:
+                            warnings.warn('client_id provided, but '
+                                          'client_secret missing')
```

Comment:
I added it because the code replaces the missing argument with an empty string and continues with authentication. I thought that seemed like an omission on the callers part, something they might need warned about. This could be an incorrect assumption.